crates/topos-tce-broadcast/src/double_echo/mod.rs

```diff
@@ -412,34 +263,165 @@ impl DoubleEcho {
+        from_peer: PeerId,
+        certificate_id: CertificateId,
+        ctx: Span,
+    ) {
+        let cert_delivered = self.delivered_certificates.get(&certificate_id).is_some();
+        if !cert_delivered {
+            if self.known_certificates.get(&certificate_id).is_some() {
+                let span = if let Some(root) = self.span_tracker.get(&certificate_id) {
+                    info_span!(
+                        parent: root,
+                        "RECV Inbound Ready",
+                        peer = self.local_peer_id,
+                        certificate_id = certificate_id.to_string()
+                    )
+                } else {
+                    info_span!(
+                        "RECV Inbound Ready",
+                        peer = self.local_peer_id,
+                        certificate_id = certificate_id.to_string()
+                    )
+                };
+
+                let _enter = span.enter();
+                debug!(
+                    "Handling DoubleEchoCommand::Ready from_peer: {} cert_id: {}",
+                    &from_peer, &certificate_id
+                );
 
-        self.dispatch(cert, origin);
+                self.consume_ready(from_peer, &certificate_id);
+
+                self.state_change_follow_up();
+                drop(_enter);
+                // need to deliver the certificate
+            } else if self.delivered_certificates.get(&certificate_id).is_none() {
+                // need to buffer the Ready
+                self.buffered_messages
+                    .entry(certificate_id)
+                    .or_default()
+                    .push(DoubleEchoCommand::Ready {
+                        from_peer,
+                        certificate_id,
+                        ctx,
+                    });
+                DOUBLE_ECHO_BUFFERED_MESSAGE_COUNT.inc();
+            }
+        }
     }
 
-    pub(crate) fn handle_deliver(&mut self, cert: Certificate) {
-        self.dispatch(cert, false)
+    pub(crate) fn consume_echo(&mut self, from_peer: PeerId, certificate_id: &CertificateId) {
+        if let Some((_certificate, state)) = self.cert_candidate.get_mut(certificate_id) {
+            Self::sample_consume_peer(&from_peer, state, SampleType::EchoSubscription);
+        }
+    }
+
+    pub(crate) fn consume_ready(&mut self, from_peer: PeerId, certificate_id: &CertificateId) {
+        if let Some((_certificate, state)) = self.cert_candidate.get_mut(certificate_id) {
+            Self::sample_consume_peer(&from_peer, state, SampleType::ReadySubscription);
+        }
     }
 
     /// Called to process potentially new certificate:
     /// - either submitted from API ( [tce_transport::TceCommands::Broadcast] command)
     /// - or received through the gossip (first step of protocol exchange)
     #[instrument(skip_all)]
-    pub(crate) fn dispatch(&mut self, cert: Certificate, origin: bool) {
+    pub(crate) fn broadcast(&mut self, cert: Certificate, origin: bool) {
```

Comment:
Same with `broadcast`. Would maybe be helpful to have this in its own file, so one can go over this critical piece of code without distraction?

Reply:
Plan for next iteration